### DISCUSSION TITLE
新基礎3　adctestについて

### DISCUSSION THREAD

**Discussion by shunta-nambu** [toppers/users-forum#115] [TOPPERS Seminar] [answered]
竹内様
お世話になっております。南部駿太と申します。
新基礎3 1日目のadctestに関してお尋ねします。

adctestでは、AD変換の結果をDMAによってメモリへ転送するプログラムだと理解しています。
adctest.cfgにおいてAD変換の割込みとDMA割込みを有効にしていますが、正常に機器が動作している場合、”実際に発生するのはDMA割込みのみ”という理解でよろしいでしょうか？

以下に私の理解の範囲でプログラムの流れを記述します。
1．クロックやチャンネル（ジョイスティックx軸、y軸の2つ＝レギュラーグループ）などの設定する
2．10ms分の変換データを転送するように指定し、DMAモードでAD変換を開始する。
3．AD変換された値はy, x, y, x, ...と次々にDMAによって配列に格納される（DMAのインクリメント設定が有効なため）
4．指定した長さ分転送が終了したとき、DMA割込みからコールバック関数を呼び出し、値を表示

DMAモードでは、AD変換完了後は割込みではなくDMAにリクエスト→転送、と理解しました。
間違えている点などがあれば、ご教示いただけますと幸いです。
よろしくお願いします。

**Comment by Ryohsuke-Takeuchi** [comment, marked as answer]
その理解で正しいです。
adc_start_dma関数では、ADCの割込み設定はオーバーラン割込みのみを設定しており、ADCの割込みはエラー発生時のみで
データ転送終了は、DMAの完了割込みとなります。
ちなみに、adc_start_int関数は、オーバーラン割込みとEOCの割込みを設定しており、ADC変換終了時も割込みが発生します。
通常の機器では、ADCは10ms単位にADC変換を行って、５回位の平均値をとるような使い方が多いようです。PICやArduinoではポーリングでADC変換待ちを行っているようです。リアルタイムカーネルを使う場合は、ADC変換終了待ちでCPUを占有されるのはいやなので、タスクでキックして割込みで取り込むような手法が良いと思います。

**Reply by shunta-nambu** [reply]
ご回答ありがとうございます。理解が深まりました。
位置取り込みプログラム (joy)では、ADC変換完了時の割込みが発生していることを確認しました。
引き続きよろしくお願いいたします。
